plugin/documents.py

```diff
@@ -136,9 +136,17 @@ def detach(self) -> None:
         self.ids_to_listeners.pop(self.buffer.buffer_id, None)
         super().detach()
 
-    def on_text_changed_async(self, changes: Iterable[sublime.TextChange]) -> None:
-        for listener in list(self.view_listeners):
-            listener.on_text_changed_async(changes)
+    def on_text_changed(self, changes: Iterable[sublime.TextChange]) -> None:
+        view = self.buffer.primary_view()
+        if not view:
+            return
```

Comment:
There can be a case when there is no primary view? And if so, is it really the right thing to not call the listeners then?

Reply:
Yeah, I am kind of conflicted about this.

It is logically impossible to not have a primary view inside of `on_text_changed`.
But it is possible to not have a primary view when doing `buffer.primary_view()` after a 100 second timeout with `sublime.set_timeout`.
The mypy type system cannot describe this runtime behavior accurately. Or maybe we are not using a feature of the type system that could accurately describe this.

Reply:
Essentially what we want to describe in the type system is something like this: if T is a sublime.View or sublime.Window, then any function which takes a T, becomes a function which takes an Optional[T] whenever it is passed to `sublime.set_timeout`.

Reply:
I think mypy uses such system in relation to lambdas and variables that those are capturing in their scope. But not sure now in which cases exactly.